File: frontend/app/lib/shared/models/user_data.dart
/// User data models — mirrors Web SearchHistoryItem, PlayHistoryItem, FavoriteItem, etc.
library;

class SearchHistoryItem {
  final int id;
  final String keyword;
  final int updatedAt;
  const SearchHistoryItem({required this.id, required this.keyword, required this.updatedAt});
  factory SearchHistoryItem.fromJson(Map<String, dynamic> j) => SearchHistoryItem(
    id: j['id'] as int? ?? 0,
    keyword: j['keyword'] as String? ?? '',
    updatedAt: j['updated_at'] as int? ?? 0,
  );
}

class PlayHistoryItem {
  final int id;
  final int vodId;
  final String vodName;
  final String vodPic;
  final String resourceDomain;
  final String resourceName;
  final String groupKey;
  final int sourceIndex;
  final int epIndex;
  final String epName;
  /// 播放进度百分比 (0-100)，与后端/Web 端一致
  final double progress;
  final double currentTime;
  final double duration;
  final int createdAt;
  final int updatedAt;

  const PlayHistoryItem({
    required this.id,
    required this.vodId,
    required this.vodName,
    required this.vodPic,
    required this.resourceDomain,
    required this.resourceName,
    required this.groupKey,
    required this.sourceIndex,
    required this.epIndex,
    required this.epName,
    required this.progress,
    required this.currentTime,
    required this.duration,
    required this.createdAt,
    required this.updatedAt,
  });

  factory PlayHistoryItem.fromJson(Map<String, dynamic> j) => PlayHistoryItem(
    id: (j['id'] as num?)?.toInt() ?? 0,
    vodId: (j['vod_id'] as num?)?.toInt() ?? 0,
    vodName: j['vod_name'] as String? ?? '',
    vodPic: j['vod_pic'] as String? ?? '',
    resourceDomain: j['resource_domain'] as String? ?? '',
    resourceName: j['resource_name'] as String? ?? '',
    groupKey: j['group_key'] as String? ?? '',
    sourceIndex: (j['source_index'] as num?)?.toInt() ?? 0,
    epIndex: (j['ep_index'] as num?)?.toInt() ?? 0,
    epName: j['ep_name'] as String? ?? '',
    progress: (j['progress'] as num?)?.toDouble() ?? 0.0,
    currentTime: (j['current_time'] as num?)?.toDouble() ?? 0.0,
    duration: (j['duration'] as num?)?.toDouble() ?? 0.0,
    createdAt: (j['created_at'] as num?)?.toInt() ?? 0,
    updatedAt: (j['updated_at'] as num?)?.toInt() ?? 0,
  );
}

class PlayHistoryListResponse {
  final int total;
  final List<PlayHistoryItem> items;
  const PlayHistoryListResponse({required this.total, required this.items});
  factory PlayHistoryListResponse.fromJson(Map<String, dynamic> j) => PlayHistoryListResponse(
    total: j['total'] as int? ?? 0,
    items: (j['items'] as List<dynamic>?)
            ?.map((e) => PlayHistoryItem.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
  );
}

class FavoriteItem {
  final int id;
  final int vodId;
  final String vodName;
  final String vodPic;
  final String doubanId;
  final String groupKey;
  final String site;
  final String resourceDomain;
  final String resourceName;
  final int createdAt;

  const FavoriteItem({
    required this.id,
    required this.vodId,
    required this.vodName,
    required this.vodPic,
    required this.doubanId,
    required this.groupKey,
    required this.site,
    required this.resourceDomain,
    required this.resourceName,
    required this.createdAt,
  });

  factory FavoriteItem.fromJson(Map<String, dynamic> j) => FavoriteItem(
    id: j['id'] as int? ?? 0,
    vodId: j['vod_id'] as int? ?? 0,
    vodName: j['vod_name'] as String? ?? '',
    vodPic: j['vod_pic'] as String? ?? '',
    doubanId: j['douban_id']?.toString() ?? '',
    groupKey: j['group_key'] as String? ?? '',
    site: j['site'] as String? ?? '',
    resourceDomain: j['resource_domain'] as String? ?? '',
    resourceName: j['resource_name'] as String? ?? '',
    createdAt: j['created_at'] as int? ?? 0,
  );
}

class FavoriteListResponse {
  final int total;
  final List<FavoriteItem> items;
  const FavoriteListResponse({required this.total, required this.items});
  factory FavoriteListResponse.fromJson(Map<String, dynamic> j) => FavoriteListResponse(
    total: j['total'] as int? ?? 0,
    items: (j['items'] as List<dynamic>?)
            ?.map((e) => FavoriteItem.fromJson(e as Map<String, dynamic>))
            .toList() ??
        [],
  );
}

class FavoriteCheckResponse {
  final bool isFavorite;
  const FavoriteCheckResponse({required this.isFavorite});
  factory FavoriteCheckResponse.fromJson(Map<String, dynamic> j) => FavoriteCheckResponse(
    isFavorite: j['is_favorite'] as bool? ?? false,
  );
}

File: frontend/web/src/api/user-data.ts
import { API } from '@/constants/api';
import { post } from './client';
import type {
  SearchHistoryItem,
  PlayHistoryItem,
  PlayHistoryListResp,
  FavoriteItem,
  FavoriteListResp,
  FavoriteCheckResp,
  PlayHistoryGetReq,
} from '@/types/api';

// ── 搜索历史 ────────────────────────────────────────────────────────────────

/** 获取搜索历史列表 */
export function getSearchHistoryList(limit?: number): Promise<SearchHistoryItem[]> {
  return post<SearchHistoryItem[]>(API.SEARCH_HISTORY_LIST, { limit: limit ?? 20 });
}

/** 新增搜索记录 */
export function addSearchHistory(keyword: string): Promise<SearchHistoryItem> {
  return post<SearchHistoryItem>(API.SEARCH_HISTORY_ADD, { keyword });
}

/** 删除单条搜索记录 */
export function deleteSearchHistory(id: number): Promise<void> {
  return post(API.SEARCH_HISTORY_DELETE, { id });
}

/** 清空搜索历史 */
export function clearSearchHistory(): Promise<void> {
  return post(API.SEARCH_HISTORY_CLEAR);
}

// ── 播放历史 ────────────────────────────────────────────────────────────────

/** 查询单条播放记录（记录不存在返回 null） */
export function getPlayHistory(params: PlayHistoryGetReq): Promise<PlayHistoryItem | null> {
  return post<PlayHistoryItem | null>(API.PLAY_HISTORY_GET, params as unknown as Record<string, unknown>);
}

/** 获取播放历史列表 */
export function getPlayHistoryList(limit?: number, offset?: number): Promise<PlayHistoryListResp> {
  return post<PlayHistoryListResp>(API.PLAY_HISTORY_LIST, {
    limit: limit ?? 20,
    offset: offset ?? 0,
  });
}

/** 新增/更新播放记录（完整 upsert） */
export function upsertPlayHistory(record: {
  vod_id: number;
  vod_name: string;
  vod_pic: string;
  resource_domain: string;
  resource_name: string;
  group_key: string;
  source_index: number;
  ep_index: number;
  ep_name: string;
  progress: number;
  current_time: number;
  duration: number;
}): Promise<PlayHistoryItem> {
  return post<PlayHistoryItem>(API.PLAY_HISTORY_UPSERT, record as unknown as Record<string, unknown>);
}

/** 更新播放进度（upsert 语义：记录不存在时自动创建） */
export function updatePlayProgress(params: {
  vod_id: number;
  vod_name: string;
  vod_pic: string;
  resource_domain: string;
  resource_name: string;
  group_key: string;
  source_index: number;
  ep_index: number;
  ep_name: string;
  progress: number;
  current_time: number;
  duration: number;
  // 流调度相关字段（可选，代理开启时附带）
  session?: string;
  current_index?: number;
  buffered_ahead?: number;
}): Promise<void> {
  return post(API.PLAY_HISTORY_PROGRESS, params as unknown as Record<string, unknown>);
}

/** 删除单条播放记录 */
export function deletePlayHistory(id: number): Promise<void> {
  return post(API.PLAY_HISTORY_DELETE, { id });
}

/** 清空播放历史 */
export function clearPlayHistory(): Promise<void> {
  return post(API.PLAY_HISTORY_CLEAR);
}

// ── 收藏 ────────────────────────────────────────────────────────────────────

/** 获取收藏列表 */
export function getFavoritesList(limit?: number, offset?: number, keyword?: string): Promise<FavoriteListResp> {
  return post<FavoriteListResp>(API.FAVORITES_LIST, {
    limit: limit ?? 50,
    offset: offset ?? 0,
    keyword: keyword ?? "",
  });
}

/** 添加收藏 */
export function addFavorite(record: {
  vod_id: number;
  vod_name: string;
  vod_pic: string;
  douban_id: string;
  group_key: string;
  site: string;
  resource_domain: string;
  resource_name: string;
}): Promise<FavoriteItem> {
  return post<FavoriteItem>(API.FAVORITES_ADD, record as unknown as Record<string, unknown>);
}

/** 取消收藏 */
export function removeFavorite(params: {
  vod_id: number;
  resource_domain: string;
  douban_id: string;
}): Promise<void> {
  return post(API.FAVORITES_REMOVE, params);
}

/** 切换收藏状态 */
export function toggleFavorite(record: {
  vod_id: number;
  vod_name: string;
  vod_pic: string;
  douban_id: string;
  group_key: string;
  site: string;
  resource_domain: string;
  resource_name: string;
}): Promise<FavoriteCheckResp> {
  return post<FavoriteCheckResp>(API.FAVORITES_TOGGLE, record as unknown as Record<string, unknown>);
}

/** 检查是否已收藏 */
export function checkFavorite(params: {
  vod_id: number;
  resource_domain: string;
  douban_id: string;
}): Promise<FavoriteCheckResp> {
  return post<FavoriteCheckResp>(API.FAVORITES_CHECK, params);
}

/** 清空收藏 */
export function clearFavorites(): Promise<void> {
  return post(API.FAVORITES_CLEAR);
}
